### PR TITLE
419239:Automate Azure DevOps Service Connection Setup with Azure Portal

### PR DESCRIPTION
**Pull request by santosh-gh**
[AB#419239](https://dev.azure.com/defragovuk/93c65135-d16b-49d6-a191-4a5313532779/_workitems/edit/395100)

What this PR does / why we need it:
Automate the creation of ADO service connection using Federated Identity Credentials

Create federated Identity credential in app reg
Create Ado service connection using federated identity credential
Pipeline flag added for creation of the service connection
This is preferred as it does not certificate and no need to update it if it expired
This will replace the existing service connection created using client secret.

Special notes for your reviewer
Any specific actions or notes on review?

Testing
https://dev.azure.com/defragovuk/DEFRA-FFC/_build/results?buildId=632997&view=results

Checklist (please delete before completing or setting auto-complete)
[x ] Story Work items associated (not Tasks)
[x ] Successful testing run(s) link provided
[x ] Title pattern should be {work item number}: {title}
[x ] Description covers all the changes in the PR
This PR contains documentation
This PR contains tests
How does this PR make you feel:

### CODE REVIEW

File: PowerShellLibrary/Add-AdAppRegistrations.ps1 [hunk 1]
```diff
@@ -506,10 +506,50 @@ Function Add-AdAppRegistrations() {
...
+
 $homeTenantContext = Get-AzContext
 
 if ($AppRegManifestStorageAccountName -or $AppRegManifestContainerName) {
     Write-Warning "AppRegManifestStorageAccountName and AppRegManifestContainerName are no longer required.  The Secret Renewal app now stores manifest data with the app registration in the notes property"
 }
 
 Add-AdAppRegistrations -appRegJsonPath $AppRegJsonPath
+
+Add-FederatedCredential -appRegJsonPath  $AppRegJsonPath
```
Comment (rajeshkaremane) [inline PowerShellLibrary/Add-AdAppRegistrations.ps1]: add federated creds only if json has federartedCredential details

Comment (santosh-gh) [reply PowerShellLibrary/Add-AdAppRegistrations.ps1]: added pipeline variable

File: PowerShellLibrary/Add-AdAppRegistrations.ps1 [hunk 2]
```diff
@@ -506,10 +508,62 @@ Function Add-AdAppRegistrations() {
...
+
+    $apps = Get-Content -Raw -Path $appRegJsonPath | ConvertFrom-Json
+    
+    foreach ($app in $apps.applications) {
+        $appReg = Get-AzADApplication -DisplayName $app.displayName       
+
+        $federatedCredentials = Get-AzADAppFederatedCredential -ApplicationObjectId $appReg.id
+        $federatedCredentials | Select-Object -Property Name
+
+        $ficName =  $app.displayName + "-fic"
```
Comment (rajeshkaremane) [inline PowerShellLibrary/Add-AdAppRegistrations.ps1]: remove -fic

File: PowerShellLibrary/Add-AdAppRegistrations.ps1 [hunk 3]
```diff
@@ -506,10 +508,62 @@ Function Add-AdAppRegistrations() {
...
+        $issuer = "https://vstoken.dev.azure.com/" + $app.adoOrganizationId
+        $subject = "sc://defragovuk/DEFRA-FFC/" + $app.subscriptionName + "-FSC"
```
Comment (rajeshkaremane) [inline PowerShellLibrary/Add-AdAppRegistrations.ps1]: Can "defragovuk/DEFRA-FFC" this be added dynamically?

Comment (rajeshkaremane) [reply PowerShellLibrary/Add-AdAppRegistrations.ps1]: Also remove -FSC

Comment (santosh-gh) [reply PowerShellLibrary/Add-AdAppRegistrations.ps1]: updated

File: PowerShellLibrary/Add-AdAppRegistrations.ps1 [hunk 4]
```diff
@@ -4,7 +4,9 @@ Param (
     [Parameter(Mandatory = $false)]
     [string]$AppRegManifestStorageAccountName,
     [Parameter(Mandatory = $false)]
-    [string]$AppRegManifestContainerName
+    [string]$AppRegManifestContainerName,
+    [Parameter(Mandatory = $false)]
+    [bool]$federatedCredential
```
Comment (rajeshkaremane) [inline PowerShellLibrary/Add-AdAppRegistrations.ps1]: Please set default value to False

File: PowerShellLibrary/Add-AdAppRegistrations.ps1 [hunk 5]
```diff
@@ -272,29 +274,33 @@ Function Set-AadApp {
...
-        }
+
+        }        
     }
     else {
         Write-Output "Updating Application '$($app.displayName)'"
         Invoke-RestMethod -Method Patch -Headers $headers -Uri "$applicationsUri/$($application.id)" -Body $applicationBody | Out-Null
 
-        if ($app.keyVault) {
+        if ($app.keyVault -and ($federatedCredential -eq $False)) {
```
Comment (rajeshkaremane) [inline PowerShellLibrary/Add-AdAppRegistrations.ps1]: we dont need the and condition

File: PowerShellLibrary/Add-AdAppRegistrations.ps1 [hunk 6]
```diff
@@ -272,29 +274,33 @@ Function Set-AadApp {
...
         $application = Invoke-RestMethod -Method Post -Headers $headers -Uri $applicationsUri -Body $applicationBody
 
-        $servicePrincipals = Invoke-RestMethod -Method GET -Headers $headers -Uri $($servicePrincipalUri + $filter -f $($app.displayName))
-        if ($servicePrincipals.value.Length -eq 0) {
-            $spJson = @{}
-            $spJson.Add("appId", $application.appId)
-            if ($app.appRoles) {
-                $spJson.Add("appRoleAssignmentRequired", $True)
-            }
+        if($federatedCredential -eq $False)
```
Comment (rajeshkaremane) [inline PowerShellLibrary/Add-AdAppRegistrations.ps1]: Dont need this condition

File: PowerShellLibrary/Add-AdAppRegistrations.ps1 (hunk 4 above)
Comment (santosh-gh) [reply PowerShellLibrary/Add-AdAppRegistrations.ps1]: assigned default value

File: PowerShellLibrary/Add-AdAppRegistrations.ps1 (hunk 5 above)
Comment (santosh-gh) [reply PowerShellLibrary/Add-AdAppRegistrations.ps1]: removed

File: PowerShellLibrary/Add-AdAppRegistrations.ps1 (hunk 6 above)
Comment (santosh-gh) [reply PowerShellLibrary/Add-AdAppRegistrations.ps1]: removed

**Review by rajeshkaremane — DISMISSED** [review]

**Review by rajeshkaremane — DISMISSED** [review]

**Pull request closed without merging**